### PR TITLE
Integrate Transformer Engine and FP8 support

### PR DESCRIPTION
**Pull request by SanjuCSudhakaran**
Integrate Transformer Engine and FP8 in optimum-habana-fork

### CODE REVIEW

**Review by vivekgoe — COMMENTED** [review]
Completed review of few files and added review comments. Let's discuss these tomorrow.

File: optimum/habana/accelerate/accelerator.py [hunk 1]
```diff
@@ -342,30 +371,8 @@ def prepare_model(self, model: torch.nn.Module, device_placement: bool = None, e
...
-        #     )
-        #     if not fp8_enabled:
-        #         logger.warn(
-        #             f"The current device has compute capability of {cuda_device_capacity} which is "
-        #             "insufficient for FP8 mixed precision training (requires a GPU Hopper/Ada Lovelace "
-        #             "or higher, compute capability of 8.9 or higher). Will use FP16 instead."
-        #         )
-        #     model.forward = fp8_autocast(enabled=fp8_enabled, fp8_recipe=fp8_recipe)(model.forward)
-
+        if self.state.is_fp8_enabled:
```
[the remark this answers is not in the document]
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/accelerator.py:374]: With DeepSpeed, the mixed precision type gets overwritten using the precision specified in the DeepSpeed config (even though fp8 training is enabled).

So in order to make things work even with that, I used a separate Boolean variable for checking whether fp8 is enabled.

**Review by vivekgoe — APPROVED** [review]
Looks good to me. Let's wait to conclude on fp8 convergence also and then we can merge this.

File: optimum/habana/peft/layer.py [hunk 2]
```diff
@@ -29,3 +31,41 @@ def GaudiLoraLayerLinearForward(self, x: torch.Tensor, *args: Any, **kwargs: Any
...
+class LoRALinear:
+    def __init__(self, module):
+        has_bias = module.bias is not None
+        self.module = module
+        self.module.te_linear =  te.Linear(module.in_features, module.out_features, bias=has_bias, params_dtype=module.weight.dtype, skip_weight_param_allocation=True)
+
+    def _linear(self, input: torch.Tensor) -> torch.Tensor:
+        return self.module.te_linear(input, transpose(self.module.weight, self.module.fan_in_fan_out), bias=self.module.bias)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
```
Comment (bhargaveede) [inline optimum/habana/peft/layer.py:44]: Better to align the function definition to forward(self, x: torch.Tensor, *args: Any, **kwargs: Any) -> torch.Tensor:

https://github.com/HabanaAI/optimum-habana-fork/pull/91/files#diff-3491044d2e0fd212ccdd50851ff639d6f0b4fba7d58f38bc92de53b5e1f56713R8

Comment (bhargaveede) [reply optimum/habana/peft/layer.py:44]: Found it. GaudiLoraLayerLinearForward is based on old version.

**Review by bhargaveede — APPROVED** [review]

File: optimum/habana/accelerate/accelerator.py [hunk 3]
```diff
@@ -289,6 +307,17 @@ def __init__(
...
         raise ValueError("fp16 is not supported on Habana Gaudi.")
 
+    def wrap_fp8(self, model):
+        if not has_transformer_engine_layers(model):
+            with torch.no_grad():
+                convert_model(model)
+            model._converted_to_transformer_engine = True
+        model._original_forward = model.forward
+        model.forward = te.fp8_autocast(enabled=True, fp8_recipe=self.fp8_recipe_handler)(model.forward)
+        print(f'fp8 model {model}')
```
[the remark this answers is not in the document]
Comment (SanjuCSudhakaran) [reply optimum/habana/accelerate/accelerator.py]: Done

**Pull request merged**